src/perfEvents_linux.cpp

```diff
@@ -150,6 +154,24 @@ static void adjustFDLimit() {
     }
 }
 
+static bool is_kernel_gte_3_14() {
+    struct utsname uts;
+
+    if (uname(&uts) != 0) {
+        return false;
+    }
+
+    int major, minor;
+    if (sscanf(uts.release, "%d.%d", &major, &minor) != 2) {
+        return false;
+    }
+
+    return (major > 3 || (major == 3 && minor >= 14));
+}
+
+// Useful to close perf_event fds on exec to avoid issues in child processes.
+// Skip if not supported.
+static int CLOEXEC_FLAG = is_kernel_gte_3_14() ? PERF_FLAG_FD_CLOEXEC : 0;
```

Comment:
I tested perf_events on CentOS 7.9 (Linux 3.10.0-1160.119.1.el7.x86_64) and found that `PERF_FLAG_FD_CLOEXEC` is actually supported. It's not uncommon that some features/options are backported to older kernels.
This means, relying solely on kernel major.minor version is not very accurate. Let's first call `perf_event_open` with the flag, and if the call fails, retry it without the flag. That's also what [perf](https://github.com/torvalds/linux/blob/2a520073e74fbb956b5564818fc5529dcc7e9f0e/tools/perf/util/cloexec.c#L59) and [other profilers](https://github.com/elastic/go-perf/blob/master/perf.go#L210-L211) do.

For the sake of simplicity, `check()` and `supported()` may call `perf_event_open` just once without `flags == 0`.

Reply:
Thanks, updated. Would you like me to use fcntl directly on the fd if PERF_FLAG_FD_CLOEXEC is not supported, or just ignore that case?
I also didn't cache the supported/not supported decision - it should be widely supported and if not it just means we'll double the number of perf_event_open calls, which doesn't seem too bad, but let me know if you'd prefer to cache it.

Reply:
Thanks!